rust/blockstore/src/arrow/block/delta/ordered_merge_delta.rs

```diff
@@ -0,0 +1,227 @@
+use crate::{
+    arrow::{
+        block::Block,
+        types::{
+            ArrowReadableKey, ArrowReadableValue, ArrowWriteableKey, ArrowWriteableValue,
+            BuilderMutationOrderHint,
+        },
+    },
+    key::{CompositeKey, KeyWrapper},
+};
+use arrow::array::{RecordBatch, StringArray};
+use uuid::Uuid;
+
+/// A block delta tracks a source block and represents the new state of a block. Blocks are
+/// immutable, so when a write is made to a block, a new block is created with the new state.
+/// A block delta is a temporary representation of the new state of a block. A block delta
+/// can be converted to a block data, which is then used to create a new block. A block data
+/// can be converted into a block delta for new writes.
+/// # Methods
+/// - add: adds a key value pair to the block delta.
+/// - delete: deletes a key from the block delta.
+/// - get_size: gets the size of the block delta.
+/// - split: splits the block delta into new block deltas based on a max block size.
+pub struct OrderedMergeBlockDelta {
+    pub(in crate::arrow) builder: BlockStorage,
+    pub(in crate::arrow) id: Uuid,
+    copied_up_to_row_of_old_block: usize,
+    old_block: Option<Block>,
+}
+
+impl DeltaCommon for OrderedMergeBlockDelta {
+    // NOTE(rescrv):  K is unused, but it is very conceptually easy to think of everything as
+    // key-value pairs.  I started to refactor this to remove ArrowWriteableKey, but it was not
+    // readable to tell whether I was operating on the key or value type.  Keeping both but
+    // suppressing the clippy error is a reasonable alternative.
+    #[allow(clippy::extra_unused_type_parameters)]
+    fn new<K: ArrowWriteableKey, V: ArrowWriteableValue>(id: Uuid) -> Self {
+        OrderedMergeBlockDelta {
+            builder: V::get_delta_builder(BuilderMutationOrderHint::Ordered),
+            id,
+            copied_up_to_row_of_old_block: 0,
+            old_block: None,
+        }
+    }
+
+    fn fork_block<K: ArrowWriteableKey, V: ArrowWriteableValue>(id: Uuid, block: &Block) -> Self {
+        let mut delta = OrderedMergeBlockDelta::new::<K, V>(id);
+        // todo: is this clone cheap?
+        delta.old_block = Some(block.clone());
+        delta
+    }
+
+    fn id(&self) -> Uuid {
+        self.id
+    }
+
+    fn finish<K: ArrowWriteableKey, V: ArrowWriteableValue>(
+        self,
+        metadata: Option<HashMap<String, String>>,
+    ) -> RecordBatch {
+        self.builder.into_record_batch::<K>(metadata)
+    }
+}
+
+impl OrderedMergeBlockDelta {
+    pub fn add<K, V>(&mut self, prefix: &str, key: K, value: V)
+    where
+        K: ArrowWriteableKey,
+        V: ArrowWriteableValue,
+    {
+        if let Some(old_block) = self.old_block.clone().as_ref() {
+            // todo: is clone expensive here?
+            self.copy_up_to::<K::ReadableKey<'_>, V::ReadableValue<'_>>(
+                prefix,
+                key.clone().into(),
```

Comment:
what is a good way to avoid cloning/transforming the key here? I was thinking we could introduce a new `BorrowedKeyWrapper` that implements `AsRef<ArrowWriteableKey> + AsRef<ArrowReadableKey>` but maybe there's a cleaner solution